diff --git a/Modules/_collectionsmodule.c b/Modules/_collectionsmodule.c
index e96a546a818d3d..c2e8741735b3f8 100644
--- a/Modules/_collectionsmodule.c
+++ b/Modules/_collectionsmodule.c
@@ -746,9 +746,10 @@ deque_clear(PyObject *self)
        adversary could cause it to never terminate).
     */
 
+    PyObject *old_exc = PyErr_GetRaisedException();
     b = newblock(deque);
+    PyErr_SetRaisedException(old_exc);
     if (b == NULL) {
-        PyErr_Clear();
         goto alternate_method;
     }
 
