diff --git a/Python/executor_cases.c.h b/Python/executor_cases.c.h
index 3d27fd67090..06168a1c88a 100644
--- a/Python/executor_cases.c.h
+++ b/Python/executor_cases.c.h
@@ -8193,13 +8193,25 @@
             oparg = CURRENT_OPARG();
             v = _stack_item_0;
             list = stack_pointer[-1 - (oparg-1)];
+            stack_pointer[0] = v;
+            stack_pointer += 1;
+            ASSERT_WITHIN_STACK_BOUNDS(__FILE__, __LINE__);
+            _PyFrame_SetStackPointer(frame, stack_pointer);
             int err = _PyList_AppendTakeRef((PyListObject *)PyStackRef_AsPyObjectBorrow(list),
                 PyStackRef_AsPyObjectSteal(v));
+            stack_pointer = _PyFrame_GetStackPointer(frame);
             if (err < 0) {
+                stack_pointer += -1;
+                ASSERT_WITHIN_STACK_BOUNDS(__FILE__, __LINE__);
                 SET_CURRENT_CACHED_VALUES(0);
                 JUMP_TO_ERROR();
             }
+            _tos_cache0 = PyStackRef_ZERO_BITS;
+            _tos_cache1 = PyStackRef_ZERO_BITS;
+            _tos_cache2 = PyStackRef_ZERO_BITS;
             SET_CURRENT_CACHED_VALUES(0);
+            stack_pointer += -1;
+            ASSERT_WITHIN_STACK_BOUNDS(__FILE__, __LINE__);
             ASSERT_WITHIN_STACK_BOUNDS_IGNORING_CACHE(__FILE__, __LINE__);
             break;
         }
@@ -18861,137 +18873,6 @@
             break;
         }
 
-        case _CALL_LIST_APPEND_r03: {
-            CHECK_CURRENT_CACHED_VALUES(0);
-            ASSERT_WITHIN_STACK_BOUNDS_IGNORING_CACHE(__FILE__, __LINE__);
-            _PyStackRef arg;
-            _PyStackRef self;
-            _PyStackRef callable;
-            _PyStackRef none;
-            _PyStackRef c;
-            _PyStackRef s;
-            oparg = CURRENT_OPARG();
-            arg = stack_pointer[-1];
-            self = stack_pointer[-2];
-            callable = stack_pointer[-3];
-            assert(oparg == 1);
-            PyObject *self_o = PyStackRef_AsPyObjectBorrow(self);
-            if (!LOCK_OBJECT(self_o)) {
-                UOP_STAT_INC(uopcode, miss);
-                SET_CURRENT_CACHED_VALUES(0);
-                JUMP_TO_JUMP_TARGET();
-            }
-            STAT_INC(CALL, hit);
-            int err = _PyList_AppendTakeRef((PyListObject *)self_o, PyStackRef_AsPyObjectSteal(arg));
-            UNLOCK_OBJECT(self_o);
-            if (err) {
-                SET_CURRENT_CACHED_VALUES(0);
-                JUMP_TO_ERROR();
-            }
-            c = callable;
-            s = self;
-            none = PyStackRef_None;
-            _tos_cache2 = s;
-            _tos_cache1 = c;
-            _tos_cache0 = none;
-            SET_CURRENT_CACHED_VALUES(3);
-            stack_pointer += -3;
-            ASSERT_WITHIN_STACK_BOUNDS(__FILE__, __LINE__);
-            ASSERT_WITHIN_STACK_BOUNDS_IGNORING_CACHE(__FILE__, __LINE__);
-            break;
-        }
-
-        case _CALL_LIST_APPEND_r13: {
-            CHECK_CURRENT_CACHED_VALUES(1);
-            ASSERT_WITHIN_STACK_BOUNDS_IGNORING_CACHE(__FILE__, __LINE__);
-            _PyStackRef arg;
-            _PyStackRef self;
-            _PyStackRef callable;
-            _PyStackRef none;
-            _PyStackRef c;
-            _PyStackRef s;
-            _PyStackRef _stack_item_0 = _tos_cache0;
-            oparg = CURRENT_OPARG();
-            arg = _stack_item_0;
-            self = stack_pointer[-1];
-            callable = stack_pointer[-2];
-            assert(oparg == 1);
-            PyObject *self_o = PyStackRef_AsPyObjectBorrow(self);
-            if (!LOCK_OBJECT(self_o)) {
-                UOP_STAT_INC(uopcode, miss);
-                _tos_cache0 = arg;
-                SET_CURRENT_CACHED_VALUES(1);
-                JUMP_TO_JUMP_TARGET();
-            }
-            STAT_INC(CALL, hit);
-            int err = _PyList_AppendTakeRef((PyListObject *)self_o, PyStackRef_AsPyObjectSteal(arg));
-            UNLOCK_OBJECT(self_o);
-            if (err) {
-                stack_pointer += 1;
-                ASSERT_WITHIN_STACK_BOUNDS(__FILE__, __LINE__);
-                SET_CURRENT_CACHED_VALUES(0);
-                JUMP_TO_ERROR();
-            }
-            c = callable;
-            s = self;
-            none = PyStackRef_None;
-            _tos_cache2 = s;
-            _tos_cache1 = c;
-            _tos_cache0 = none;
-            SET_CURRENT_CACHED_VALUES(3);
-            stack_pointer += -2;
-            ASSERT_WITHIN_STACK_BOUNDS(__FILE__, __LINE__);
-            ASSERT_WITHIN_STACK_BOUNDS_IGNORING_CACHE(__FILE__, __LINE__);
-            break;
-        }
-
-        case _CALL_LIST_APPEND_r23: {
-            CHECK_CURRENT_CACHED_VALUES(2);
-            ASSERT_WITHIN_STACK_BOUNDS_IGNORING_CACHE(__FILE__, __LINE__);
-            _PyStackRef arg;
-            _PyStackRef self;
-            _PyStackRef callable;
-            _PyStackRef none;
-            _PyStackRef c;
-            _PyStackRef s;
-            _PyStackRef _stack_item_0 = _tos_cache0;
-            _PyStackRef _stack_item_1 = _tos_cache1;
-            oparg = CURRENT_OPARG();
-            arg = _stack_item_1;
-            self = _stack_item_0;
-            callable = stack_pointer[-1];
-            assert(oparg == 1);
-            PyObject *self_o = PyStackRef_AsPyObjectBorrow(self);
-            if (!LOCK_OBJECT(self_o)) {
-                UOP_STAT_INC(uopcode, miss);
-                _tos_cache1 = arg;
-                _tos_cache0 = self;
-                SET_CURRENT_CACHED_VALUES(2);
-                JUMP_TO_JUMP_TARGET();
-            }
-            STAT_INC(CALL, hit);
-            int err = _PyList_AppendTakeRef((PyListObject *)self_o, PyStackRef_AsPyObjectSteal(arg));
-            UNLOCK_OBJECT(self_o);
-            if (err) {
-                stack_pointer[0] = self;
-                stack_pointer += 2;
-                ASSERT_WITHIN_STACK_BOUNDS(__FILE__, __LINE__);
-                SET_CURRENT_CACHED_VALUES(0);
-                JUMP_TO_ERROR();
-            }
-            c = callable;
-            s = self;
-            none = PyStackRef_None;
-            _tos_cache2 = s;
-            _tos_cache1 = c;
-            _tos_cache0 = none;
-            SET_CURRENT_CACHED_VALUES(3);
-            stack_pointer += -1;
-            ASSERT_WITHIN_STACK_BOUNDS(__FILE__, __LINE__);
-            ASSERT_WITHIN_STACK_BOUNDS_IGNORING_CACHE(__FILE__, __LINE__);
-            break;
-        }
-
         case _CALL_LIST_APPEND_r33: {
             CHECK_CURRENT_CACHED_VALUES(3);
             ASSERT_WITHIN_STACK_BOUNDS_IGNORING_CACHE(__FILE__, __LINE__);
@@ -19019,13 +18900,16 @@
                 JUMP_TO_JUMP_TARGET();
             }
             STAT_INC(CALL, hit);
+            stack_pointer[0] = callable;
+            stack_pointer[1] = self;
+            stack_pointer[2] = arg;
+            stack_pointer += 3;
+            ASSERT_WITHIN_STACK_BOUNDS(__FILE__, __LINE__);
+            _PyFrame_SetStackPointer(frame, stack_pointer);
             int err = _PyList_AppendTakeRef((PyListObject *)self_o, PyStackRef_AsPyObjectSteal(arg));
+            stack_pointer = _PyFrame_GetStackPointer(frame);
             UNLOCK_OBJECT(self_o);
             if (err) {
-                stack_pointer[0] = callable;
-                stack_pointer[1] = self;
-                stack_pointer += 3;
-                ASSERT_WITHIN_STACK_BOUNDS(__FILE__, __LINE__);
                 SET_CURRENT_CACHED_VALUES(0);
                 JUMP_TO_ERROR();
             }
@@ -19036,6 +18920,8 @@
             _tos_cache1 = c;
             _tos_cache0 = none;
             SET_CURRENT_CACHED_VALUES(3);
+            stack_pointer += -3;
+            ASSERT_WITHIN_STACK_BOUNDS(__FILE__, __LINE__);
             ASSERT_WITHIN_STACK_BOUNDS_IGNORING_CACHE(__FILE__, __LINE__);
             break;
         }
diff --git a/Python/generated_cases.c.h b/Python/generated_cases.c.h
index 94384d5db3c..cfd8e5b282d 100644
--- a/Python/generated_cases.c.h
+++ b/Python/generated_cases.c.h
@@ -3809,7 +3809,9 @@
                     JUMP_TO_PREDICTED(CALL);
                 }
                 STAT_INC(CALL, hit);
+                _PyFrame_SetStackPointer(frame, stack_pointer);
                 int err = _PyList_AppendTakeRef((PyListObject *)self_o, PyStackRef_AsPyObjectSteal(arg));
+                stack_pointer = _PyFrame_GetStackPointer(frame);
                 UNLOCK_OBJECT(self_o);
                 if (err) {
                     JUMP_TO_LABEL(error);
@@ -8202,8 +8204,10 @@
             _PyStackRef v;
             v = stack_pointer[-1];
             list = stack_pointer[-2 - (oparg-1)];
+            _PyFrame_SetStackPointer(frame, stack_pointer);
             int err = _PyList_AppendTakeRef((PyListObject *)PyStackRef_AsPyObjectBorrow(list),
                 PyStackRef_AsPyObjectSteal(v));
+            stack_pointer = _PyFrame_GetStackPointer(frame);
             if (err < 0) {
                 JUMP_TO_LABEL(pop_1_error);
             }
diff --git a/Tools/cases_generator/analyzer.py b/Tools/cases_generator/analyzer.py
index 22a321b4953..ea58a37e07e 100644
--- a/Tools/cases_generator/analyzer.py
+++ b/Tools/cases_generator/analyzer.py
@@ -661,7 +661,6 @@ def has_error_without_pop(op: parser.CodeDef) -> bool:
     "_PyGen_GetGeneratorFromFrame",
     "gen_try_set_executing",
     "_PyInterpreterState_GET",
-    "_PyList_AppendTakeRef",
     "_PyList_ITEMS",
     "_PyLong_CompactValue",
     "_PyLong_DigitCount",
