diff --git a/include/forkserver.h b/include/forkserver.h
index a6f62cf4..4c5d1fda 100644
--- a/include/forkserver.h
+++ b/include/forkserver.h
@@ -282,6 +282,8 @@ typedef enum fsrv_run_result {
 
 void afl_fsrv_init(afl_forkserver_t *fsrv);
 void afl_fsrv_init_dup(afl_forkserver_t *fsrv_to, afl_forkserver_t *from);
+const char *afl_fsrv_run_result_name(afl_forkserver_t *fsrv,
+                                     fsrv_run_result_t result);
 void afl_fsrv_setup_preload(afl_forkserver_t *fsrv, char *argv0);
 void afl_fsrv_start(afl_forkserver_t *fsrv, char **argv,
                     volatile u8 *stop_soon_p, u8 debug_child_output);
diff --git a/src/afl-forkserver.c b/src/afl-forkserver.c
index 33e9cde3..069703db 100644
--- a/src/afl-forkserver.c
+++ b/src/afl-forkserver.c
@@ -606,6 +606,60 @@ void afl_fsrv_init_dup(afl_forkserver_t *fsrv_to, afl_forkserver_t *from) {
 
 }
 
+const char *afl_fsrv_run_result_name(afl_forkserver_t *fsrv,
+                                     fsrv_run_result_t result) {
+
+  switch (result) {
+
+    case FSRV_RUN_OK:
+      return "OK";
+
+    case FSRV_RUN_TMOUT:
+      return "TMOUT";
+
+    case FSRV_RUN_CRASH:
+      if (fsrv && WIFEXITED(fsrv->child_status)) {
+
+        if ((fsrv->uses_asan & 4) &&
+            WEXITSTATUS(fsrv->child_status) == MSAN_ERROR) {
+
+          return "MSAN";
+
+        }
+
+        if ((fsrv->uses_asan & 2) &&
+            WEXITSTATUS(fsrv->child_status) == LSAN_ERROR) {
+
+          return "LSAN";
+
+        }
+
+        if (fsrv->uses_crash_exitcode &&
+            WEXITSTATUS(fsrv->child_status) == fsrv->crash_exitcode) {
+
+          return "EXITCODE";
+
+        }
+
+      }
+
+      return "CRASH";
+
+    case FSRV_RUN_ERROR:
+      return "ERROR";
+
+    case FSRV_RUN_NOINST:
+      return "NOINST";
+
+    case FSRV_RUN_NOBITS:
+      return "NOBITS";
+
+  }
+
+  return "UNKNOWN";
+
+}
+
 void afl_fsrv_setup_preload(afl_forkserver_t *fsrv, char *argv0) {
 
   // afl-qemu-trace takes care of converting AFL_PRELOAD
@@ -3005,6 +3059,11 @@ store_persistent_record: {
         ck_write(fd, data, len, fn);
         close(fd);
 
+        fprintf(stderr,
+                "[RECORD] child pid: %d, result: %s, status:  %d, file: %s\n",
+                fsrv->last_child_pid, afl_fsrv_run_result_name(fsrv, retval),
+                fsrv->child_status, fn);
+
       }
 
     }
diff --git a/src/afl-fuzz-bitmap.c b/src/afl-fuzz-bitmap.c
index 65225c9a..0f2bab26 100644
--- a/src/afl-fuzz-bitmap.c
+++ b/src/afl-fuzz-bitmap.c
@@ -1301,6 +1301,12 @@ may_save_fault:
     ck_write(fd, mem, len, fn);
     close(fd);
 
+    fprintf(stderr,
+            "[RECORD] child pid: %d, result: %s, status:  %d, file: %s\n",
+            afl->fsrv.last_child_pid,
+            afl_fsrv_run_result_name(&afl->fsrv, fault),
+            afl->fsrv.child_status, fn);
+
   }
 
   if (unlikely(afl->afl_env.afl_crash_traces) && is_crash_save) {
